### DISCUSSION TITLE
What is the Typescript type for the parameters in a server component?

### DISCUSSION THREAD

**Discussion by vedantroy** [vercel/next.js#46131] [Help]
I recently discovered that the react server components in the app directory are passed a `params` object in the props.

I was wondering, is there a Typescript type for a react server component's props?

**Comment by icyJoseph** [comment]
Not right now I think. The docs define it on spot:

- https://beta.nextjs.org/docs/api-reference/file-conventions/page

```ts
type PageProps = {
  params: { slug: string };
  searchParams?: { [key: string]: string | string[] | undefined };
}
```

One slight tricky, not impossible, but tricky, thing is that `params` would have a key bound to the path of the file (`slug` for example). I think the Next.js team is working on a type safe router as well, so perhaps it can be done by the framework.

Since an specific type or interface, is not documented at the moment, I'd just take it from the code snippet in https://beta.nextjs.org/docs/api-reference/file-conventions/page

**Comment by felipemullen** [comment]
Would love to have a reliable type that comes from `next` imports to assign to my page props. Hopefully this doesn't get forgotten by the team

**Comment by Irrelon** [comment]
IMHO it's worth making it a generic so you can specify either a string or string[] for slugs because you might want to use catch-all routes:

```typescript
export interface NextPageProps<SlugType = string> {
	params: { slug: SlugType };
	searchParams?: { [key: string]: string | string[] | undefined };
}
```

**Comment by kasir-barati** [comment]
# My temporary solution

## First defined a generic interface:
```ts
export interface ServerSideComponentProp<
    Params,
    SearchParams = undefined,
> {
    params: Params;
    searchParams: SearchParams;
}
```

## Then in my server side component:
```ts
// ./app/dashboard/[dashboardId]/page.tsx
import { ServerSideComponentProp } from './server-side-components-prop';

export default function Dashboard({
    params,
}: ServerSideComponentProp<{ dashboardId: string }>) {
    return <div>{params.dashboardId}</div>;
}
```

**Comment by Simoon-F** [comment]
I'm curious, theoretically nextjs should provide it.

**Reply by danbamber** [reply]
```
import { ReadonlyURLSearchParams } from 'next/navigation';
import { Params } from 'react-router-dom';

export type ServerProps = {
    params: Params;
    searchParams:  ReadonlyURLSearchParams;
};
```

not sure if this is a more official way? It's what I've just decided to use.

**Comment by Erik35072** [comment]
> I recently discovered that the react server components in the app directory are passed a `params` object in the props.
> 
> I was wondering, is there a Typescript type for a react server component's props?

https://nextjs.org/docs/app/building-your-application/optimizing/metadata

**Reply by LassazVegaz** [reply]
what does metadata have to do with server component props?

**Reply by richardweaver** [reply]
Dynamic metadata receives the same props and there's an example of the typing in the example: 
```
type Props = {
  params: { id: string }
  searchParams: { [key: string]: string | string[] | undefined }
}
```

**Comment by bezreyhan** [comment]
I believe this is the updated type based off of the [docs](https://nextjs.org/docs/app/api-reference/file-conventions/page#params-optional):

```
export type PageProps<T extends { [key: string]: string } = undefined> = {
  params: Promise<T>;
  searchParams: Promise<{
    [key: string]: string | string[] | undefined;
  }>;
};
```

**Comment by Netail** [comment]
We use the following:

```ts
export type SearchParams = {
    [key: string]: string | string[] | undefined;
};

export type NextAppPage<T = undefined> = T extends undefined
    ? FC<{
          searchParams?: Promise<SearchParams>;
      }>
    : FC<{
          params: Promise<T>;
          searchParams?: Promise<SearchParams>;
      }>;

export type NextAppLayout<T = undefined> = T extends undefined
    ? FC<PropsWithChildren>
    : FC<
          PropsWithChildren<{
              params: Promise<T>;
          }>
      >;
```
